### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by hcpl**
- [x] Make the crate build on stable.
- [x] Refactor TL codegen to use types from `syn` as the strongly-typed lowest-level (i.e. most resemblant to resulting Rust code) intermediate representation to avoid generation of syntactically invalid code.
  * @habnabit suggested the code to be more readable, so the current approach is to parse output from `quote!` using `syn` parsing functions.
- [x] Gate `parsing` and `printing` features of `syn` into corresponding own features to keep dependencies at bare minimum (but `tl_codegen` requires the `full` feature due to using `syn` types as IR at full capabilities).
  * That was a decision which made the code less readable, so now `syn` uses all features, including `full` (it should but due to how cargo unifies features without differentiating them for different crates whatsoever, so that `mtproto` still builds while `tl_codegen` itself cannot, I forgot to remove the `parsing` feature from `tl_codegen` - will be fixed afterwards).
- [x] Leverage `serde` infrastructure which provides both an idiomatic way to do [de]serialization in Rust and potential interoperability with other data formats such as JSON.
  * implementation of the actual MTProto data format within `serde` framework is done in [`serde_mtproto`](https://github.com/hcpl/serde_mtproto).
- [x] Add authorization examples as a launching pad for asynchronous support of (potentially) CPU-heavy crypto and IO-bound requests & responses.
  * For now all authorization examples only work up to [step 4](https://core.telegram.org/mtproto/auth_key#presenting-proof-of-work-server-authentication) inclusively - server responses for all of them currently are "404 Not Found"s.

### CODE REVIEW

**Comment by JuanPotato** [comment]
Hey dude this looks amazing. If you'd like to join a group for people making mtproto libraries send me a PM I'm @JuanPotato on telegram. There's quite a few people in the group that can help out if you have issues.

File: examples/http_auth.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            Ok(v) => v,
+            Err(e) => bailf!(e),
+        }
+    }
+}
+
+
+fn auth(handle: Handle) -> Box<Future<Item = (), Error = error::Error>> {
+    let app_info = tryf!(AppInfo::read_from_toml_file("AppInfo.toml")
+        .chain_err(|| "this example needs a AppInfo.toml file with `api_id` and `api_hash` fields in it"));
```
Comment (habnabit) [inline examples/http_auth.rs]: hmm, any reason why not `dotenv`? that's what I'd been using.

File: src/rpc/utils.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    Ok(bytes)
+}
+
+
+#[derive(Debug)]
+pub enum EitherRef<'a, T: 'a> {
+    Ref(&'a T),
+    Owned(T),
+}
```
Comment (habnabit) [inline src/rpc/utils.rs:24]: isn't this `Cow`?

**Comment by habnabit** [comment]
I'm assuming this is still WIP because there's `println!` calls in it.

File: tl_codegen/src/generator.rs [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+        for (namespaces, fn_ctors) in &self.functions {
+            for fn_ctor in fn_ctors {
+                let input_data = get_input_data(&namespaces, fn_ctor);
+                all_ctors_input_data.push(input_data);
+            }
+        }
+
+        all_ctors_input_data
+    }
+}
```
Comment (habnabit) [inline tl_codegen/src/generator.rs:553]: well this sure deletes all the quoting. :(
I find it a lot less readable to construct `syn` structs instead of using `quote!`

File: examples/http_auth.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        println!("Known public key fingerprints: {:?}", known_fingerprints);
+        let server_pk_fingerprints = res_pq.server_public_key_fingerprints.inner().as_slice();
+        println!("Server public key fingerprints: {:?}", &server_pk_fingerprints);
+        let (rsa_public_key, fingerprint) =
+            tryf!(asymm::find_first_key_fail_safe(server_pk_fingerprints));
+        println!("RSA public key used: {:#?}", &rsa_public_key);
+        let encrypted_data = tryf!(rsa_public_key.encrypt(&p_q_inner_data_serialized));
+        println!("Encrypted data: {:?}", encrypted_data.as_ref());
+        let encrypted_data2 = tryf!(rsa_public_key.encrypt2(&p_q_inner_data_serialized));
+        println!("Encrypted data 2: {:?}", &encrypted_data2);
```
Comment (Connicpu) [inline examples/http_auth.rs]: I'm assuming these are just for debugging purposes. Would you like to add the `log` crate and change these to `trace!` perhaps?

**Review by Connicpu — CHANGES_REQUESTED** [review]
I'm really happy with this refactor :smile: If you could just remove the println! debug stuff, and maybe replace it with `log` crate statements, I'll be happy to merge this!

File: src/rpc/utils.rs (hunk 2 above)
Comment (hcpl) [reply src/rpc/utils.rs:24]: In short: we need a certain degree of dependency between the borrowed and the owned variants, but not too much.`Cow` is too restrictive while `A` and `B` in `Either<A, B>` are completely independent from each other, so that even `type EitherRef<'a, T> = Either<&'a T, T>;` won't help much.

Here is a very long explanation which was unconsciously typing down before the short summary above and now I don't feel like I can delete it. Sorry for being too wordy! :D
<details>
  <summary>Long text</summary>

* There is a blanket
  ```rust
  impl<T> ToOwned for T
      where T: Clone
  {
      type Owned = T;
      fn to_owned(&self) -> T {
          self.clone()
      }

      // Irrelevant for us
      // ...
  }
  ```
  implementation - if we keep using just `Cow<'a, T>`, the trait bounds will force us to require `T: Clone` everywhere just to fulfill the type correctness when the semantics of `Message` [de]serialization doesn't even need cloning anywere;
* So we need to get rid of the `Clone` bound but preserve the borrowing ability. The only safe way to do that is disallowing creating owned values from borrowed ones (correct me if I'm wrong here). Let's show that we can't do this.
  Assume that we can do it. This boils down to solving the problem:
  Find such types `U != !` and `V != !` (empty types are useless for us), so that:
  - for any value `u` of type `U`: there exists a value `v` of type `V`: so that `&u` can be borrowed as `&v`;
  - for any type `V: ToOwned`: there doesn't exist any value in `V::Owned`.

  Both points above, combined, are equivalent to
  - all of `U: Borrow<V>`, `V: ToOwned` and `V::Owned = !` must be true.

  That implies `!: Borrow<V>` as the only solution. So `U = !` as the conclusion. Contradiction with proposition.
* Okay, so how about non-`Cow` solutions? `Either` comes to me as the only one (please propose if there are more than I know of). The downside of it was the [de]serialization: it behaves like any enum with two variants with a single field in each would behave, when we need a special treatment of owned vs. borrowed values.

Unable to find a ready-to-use solution for this case (which is, I guess, rather obscure if community hasn't come up with it yet), I rolled out mine which you can see now. If you know something better, please tell me.

Thank you for reading up to this point! :D

</details>

File: examples/http_auth.rs (hunk 1 above)
Comment (hcpl) [reply examples/http_auth.rs]: @habnabit I think `dotenv` is an application-level utility since it only changes environment but doesn't directly help with initializing `AppInfo`.

For library usage I prefer `envy` for its type-safe nature, though I can still leverage `dotenv` for examples nevertheless, thanks for the idea!

File: tl_codegen/src/generator.rs (hunk 3 above)
Comment (hcpl) [reply tl_codegen/src/generator.rs:553]: In code I write, I'd rather trade readability and conciseness for robustness and predictability.  _That said_, I'd be **super** happy to know if there is something like `syn!` macro which is akin to `quote!` but returns type-safe `syn` structures. This would combine the best of both worlds.

File: examples/http_auth.rs (hunk 4 above)
Comment (hcpl) [reply examples/http_auth.rs]: Previously I thought "since this is an example, `println!` is ought to be a thing if you want to explain what is happening". But after both you and @habnabit pointed out, I guess it does make sense to use log calls: they can be switched on/off as you wish, `println!`s cannot.

**Comment by hcpl** [comment]
@JuanPotato yeah, I do have some questions regarding key exchange and constant 404s. Thank you for invitation!

@habnabit pretty much the same answer as here: https://github.com/Connicpu/mtproto-rs/pull/10#discussion_r147042026

@Connicpu challenge accepted :)

File: tl_codegen/src/generator.rs (hunk 3 above)
Comment (hcpl) [reply tl_codegen/src/generator.rs:553]: This works:
```rust
syn::parse_item(quote! {
    pub const LAYER: i32 = #layer;
}.as_str()).unwrap();
```
which is essentially the wanted `syn!` macro albeit with a cost of runtime parsing.

I'm going to replace some really obnoxious cases when pure-`syn` code can take up to 50 lines but `syn::parse_*` + `quote! {}.as_str()` one - only 5-10.

**Comment by JuanPotato** [comment]
@hcpl You haven't pm'd me about the group yet, just bringing it back up because it might be really helpful. @Connicpu You're welcome to join too (@JuanPotato is my tg)

**Review by Connicpu — APPROVED** [review]

**Comment by Connicpu** [comment]
Looks awesome :smile: Thanks!

**Comment by hcpl** [comment]
@JuanPotato ah, sorry for hanging up! I am now making the library more easy to use and comprehend so that requesting help wouldn't baffle people too much.

@Connicpu thanks for your patience and for merging! Can you make me a collaborator to ease my life (and yours, probably :smile:) in case if I have some patches ready?

**Comment by hcpl** [comment]
Friendly pinging @Connicpu about collaboration.
